docs/design/coreclr/botr/runtime-async-codegen.md

```diff
@@ -0,0 +1,133 @@
+
+
+
+The general responsibilities of the runtime-async code generator
+
+1. Wrap the body of Task and ValueTask returning functions in try/finally blocks which set/reset the `ExecutionContext` and `SynchronizationContext`.
+
+2. Allow the async thunk logic to work.
+
+3. Generate Async Debug info (Not yet described in this document)f
+
+
+
+# Identifying calls to Runtime-Async methods that can be handled by runtime-async
+
+When compiling a call to a method that might be called in the optimized fashion, recognize the following sequence.
+
+```
+call[virt] <Method>
+[ OPTIONAL ]
+{
+[ OPTIONAL - Used for ValueTask based ConfigureAwait ]
+{
+stloc X;
+ldloca X
+}
+ldc.i4.0 / ldc.i4.1
+call[virt] <ConfigureAwait> (The virt instruction is used for ConfigureAwait on a Task based runtime async function) NI_System_Threading_Tasks_Task_ConfigureAwait
+}
+call       <Await> One of the functions which matches NI_System_Runtime_CompilerServices_AsyncHelpers_Await
+```
+
+A search for this sequence is done if Method is known to be async.
+
+If the pattern is recognized, this is a PREFIX_IS_TASK_AWAIT async call, and if the flag passed to ConfigureAwait is 1, or there is no call to ConfigureAwait, it is also a PREFIX_TASK_AWAIT_CONTINUE_ON_CAPTURED_CONTEXT call. Calls to these functions will save/restore the execution context.
+
+The dispatch to these functions will save and restore the execution context only on async dispatch.
+
+If PREFIX_TASK_AWAIT_CONTINUE_ON_CAPTURED_CONTEXT, then continuation mode shall be ContinuationContextHandling::ContinueOnCapturedContext otherwise ContinuationContextHandling::ContinueOnThreadPool.
+
+
+
+# Non-optimized pattern
+
+It is also legal for code to have a simple direct usage of NI\_System\_Runtime\_CompilerServices\_AsyncHelpers\_Await or NI\_System\_Runtime\_CompilerServices\_AsyncHelpers\_AwaitAwaiter. To support this, the Await and AwaitAwaiter functions are marked as async even though they do not return a Task/ValueTask.
+
+The dispatch to these functions will save and restore the execution context only on async dispatch.
+
+The dispatch to these functions will set continuation mode to ContinuationContextHandling::None
+
+SaveAndRestoreSynchronizationContextField is disabled
+
+# Calli of an async function
+
+The dispatch to these functions will save and restore the execution context only on async dispatch.
+
+# The System.Runtime.CompilerServices.AsyncHelpers::AsyncSuspend intrinsic
+
+When encountered, triggers the function to suspend immediately, and return the passed in Continuation.
+
+# Saving and restoring of contexts
+
+Capture the execution context before the suspension, and when the function resumes, call `AsyncHelpers.RestoreExecutionContext`. The context should be stored into the Continuation. The context may be captured by calling `AsyncHelpers.CaptureExecutionContext` or it may be captured directly off of the Thread object.
```

Comment:
```suggestion
Capture the execution context before the suspension, and when the function resumes, call `AsyncHelpers.RestoreExecutionContext`. The context should be stored into the Continuation. The context may be captured by calling `AsyncHelpers.CaptureExecutionContext`.
```
(note that there is `ExecutionContext.m_isFlowSuppressed` that needs to be respected, see #122125)